azure/services/agentpools/spec.go

```diff
@@ -152,243 +154,228 @@ type AgentPoolSpec struct {
 	EnableFIPS *bool
 }
 
-// ResourceName returns the name of the agent pool.
-func (s *AgentPoolSpec) ResourceName() string {
-	return s.Name
-}
-
-// ResourceGroupName returns the name of the resource group.
-func (s *AgentPoolSpec) ResourceGroupName() string {
-	return s.ResourceGroup
-}
-
-// OwnerResourceName is a no-op for agent pools.
-func (s *AgentPoolSpec) OwnerResourceName() string {
-	return s.Cluster
-}
-
-// CustomHeaders returns custom headers to be added to the Azure API calls.
-func (s *AgentPoolSpec) CustomHeaders() map[string]string {
-	return s.Headers
+// ResourceRef implements azure.ASOResourceSpecGetter.
+func (s *AgentPoolSpec) ResourceRef() *asocontainerservicev1.ManagedClustersAgentPool {
+	return &asocontainerservicev1.ManagedClustersAgentPool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.Name,
+			Namespace: s.Namespace,
+		},
+	}
 }
 
 // Parameters returns the parameters for the agent pool.
-func (s *AgentPoolSpec) Parameters(ctx context.Context, existing interface{}) (params interface{}, err error) {
+func (s *AgentPoolSpec) Parameters(ctx context.Context, existing *asocontainerservicev1.ManagedClustersAgentPool) (params *asocontainerservicev1.ManagedClustersAgentPool, err error) {
 	_, log, done := tele.StartSpanWithLogger(ctx, "agentpools.Service.Parameters")
 	defer done()
 
+	agentPool := &asocontainerservicev1.ManagedClustersAgentPool{}
+
 	nodeLabels := s.NodeLabels
 	if existing != nil {
-		existingPool, ok := existing.(armcontainerservice.AgentPool)
-		if !ok {
-			return nil, errors.Errorf("%T is not an armcontainerservice.AgentPool", existing)
-		}
-
 		// agent pool already exists
-		ps := *existingPool.Properties.ProvisioningState
-		if ps != string(infrav1.Canceled) && ps != string(infrav1.Failed) && ps != string(infrav1.Succeeded) {
-			msg := fmt.Sprintf("Unable to update existing agent pool in non terminal state. Agent pool must be in one of the following provisioning states: Canceled, Failed, or Succeeded. Actual state: %s", ps)
-			return nil, azure.WithTransientError(errors.New(msg), 20*time.Second)
-		}
```

Comment:
ASO already blocks based on provisioning state, so we don't need to block updating the ASO resource on the same thing.